libextractor: add support for gstreamer plugin

jorsn opened the pull request:
###### Motivation for this change

The gstreamer plugin provides support for additional common
file/tagging formats like id3 tags in mp3 files. In addition, it
e.g. exposes more tags than the FLAC plugin for FLAC files.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"` (last successful run on 2020-04-28)
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size: **86.71 MB (52.8%)**
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).
  *(apart from having no maintainer as before this change)*

###### Notify maintainers

*package has no maintainer*

###### Questions

The impact on the package closure size is quite substantial. Should `gstreamerSupport` be `false` by default? *I'll leave it true but added a comment about it.*

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-in-distress/3604/15

jtojnar commented:
This will only have an effect on the binary, not the library. The codecs will need to be provided through a patch https://github.com/NixOS/nixpkgs/pull/57619

doronbehar commented on pkgs/development/libraries/libextractor/default.nix:
```diff
@@ -1,8 +1,19 @@
+{ fetchurl, stdenv, substituteAll
+, libtool, gettext, zlib, bzip2, flac, libvorbis
 , exiv2, libgsf, rpm, pkgconfig, fetchpatch
+, gstreamerSupport ? true, gst_all_1 ? null
```
@jorsn regarding your concern for closure size increase, is there a feature missing from an end user package because libextractor isn't compile with gstreamer support? If not, I think you should set this to `false` by default.

jorsn replied:
Libextractor includes a cli tool `extract` which I am using e.g. for generating git-annex metadata. I would consider this an end user package. Would you, @doronbehar?
Without gstreamer, there is e.g. no support for id3 tags in mp3 files and worse flac tag support.

jorsn replied:
Ok, I'll leave it `true` and add a comment.

jorsn committed:
libextractor: add support for gstreamer plugin

The gstreamer plugin provides support for additional common
file/tagging formats like id3 tags in mp3 files. In addition, it
e.g. exposes more tags than the FLAC plugin for FLAC files.

Increase of closure size: 86.71 MB (52.8%)

jorsn commented:
I pushed only a comment about the package closure size so people needing a small closure know how to cut the size down.

I also successfully ran `nix-shell -p nix-review --run "nix-review wip"` again.
@jtojnar, @doronbehar Would you mind posting this PR [in that discourse thread](https://discourse.nixos.org/t/prs-already-reviewed/2617)?

doronbehar commented:
Usually it's the PR writer posting there :). @jtojnar has merge permissions so let's hope he'll do it.

jorsn commented:
> Usually it's the PR writer posting there :)

According to the top post it's the reviewers, so that committers know te PR is approved and don't have to think about it.

jtojnar commented on pkgs/development/libraries/libextractor/default.nix:
```diff
@@ -28,7 +41,12 @@ stdenv.mkDerivation rec {
       excludes = [ "ChangeLog" ];
     })
-  ];
+  ] ++ stdenv.lib.optional gstreamerSupport (substituteAll {
+    src = ./gst-hardcode-plugins.patch;
+    load_gst_plugins = stdenv.lib.concatMapStrings
+      (plugin: ''gst_registry_scan_path(gst_registry_get(), "${plugin}/lib/gstreamer-1.0");'')
+      (gstPlugins gst_all_1);
+  });
```
Just one missing comment:

```suggestion
  ] ++ stdenv.lib.optionals gstreamerSupport [
    # Libraries cannot be wrapped so we need to hardcode the plug-in paths.
    (substituteAll {
      src = ./gst-hardcode-plugins.patch;
      load_gst_plugins = stdenv.lib.concatMapStrings
        (plugin: ''gst_registry_scan_path(gst_registry_get(), "${plugin}/lib/gstreamer-1.0");'')
        (gstPlugins gst_all_1);
    })
  ];
```

jorsn replied:
Thanks! Committed and force-pushed (to keep only one commit).

jtojnar commented:
@GrahamcOfBorg build libextractor

jtojnar commented:
Thanks.
